### PR TITLE
[CUDA] Add PackedMultiHeadAttention operator

### PR DESCRIPTION
**Pull request by tianleiwu**
### Description
Add new operator for MultiHeadAttention with inputs removed padding. This only supports packed QKV format.

The following will be in future pull request:
(1) Symbolic shape inference of PackedMultiHeadAttention
(2) Add cuda kernels for separated Q, K and V
(3) Script to convert a model to packing model.

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/bert/packed_multihead_attention.cc [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+  int64_t batch_size = token_offset_dims[0];
+  int64_t sequence_length = token_offset_dims[1];
+
+  int64_t v_hidden_size = hidden_size;
+  if (query_dims.size() == 4) {
+    if (key != nullptr || value != nullptr) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "Input 'key' and 'value' is expected to be empty when 'query' has 4 dimensions in packing mode");
+    }
+  } else {  // query_dims.size() == 2
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/bert/packed_multihead_attention.cc:99]: assert this? (query_dims.size() == 2)

File: onnxruntime/contrib_ops/cuda/bert/packed_multihead_attention_impl.cu [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+          output + token_count * num_heads * qk_head_size,
+          output + 2 * token_count * num_heads * qk_head_size,
+          token_count);
+    } else {
+      ORT_ENFORCE(target_format == AttentionQkvFormat::QKV_TN3H);
+    }
+  }
+}
+
+template <typename T>
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/bert/packed_multihead_attention_impl.cu:187]: maybe it's better to move to https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/contrib_ops/cuda/bert/utils.cuh

File: docs/ContribOperators.md [hunk 3]
```diff
@@ -2883,6 +2884,79 @@ This version of the operator has been available since version 1 of the 'com.micr
...
+  
+  PackedMultiHeadAttention is designed to takes in packed input, i.e., only the real tokens without padding.
+  An input as above will be packed into 3 tensors like below:
+   - query ([q0, q4, q5, q8, q9, q10, q11])
+   - key ([k0, k4, k5, k8, k9, k10, k11])
+   - value ([v0, v4, v5, v8, v9, v10, v11])
+   - token_offset: 0, 4, 5, 8, 9, 10, 11,  1*, 2*, 3*, 6*, 7*
+   - cumulative_sequence_length: 0, 1, 1+2, 1+2+4
+  
+  The query, key and value tensors contains result of hidden embedding of real tokens after input projections.
```
Comment (wangyems) [inline docs/ContribOperators.md:2905]: nit: contains -> contain

**Review by wangyems — APPROVED** [review]

File: onnxruntime/contrib_ops/cuda/bert/packed_multihead_attention_impl.cu (hunk 2 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/bert/packed_multihead_attention_impl.cu:187]: will refactor later

File: docs/ContribOperators.md (hunk 3 above)
Comment (tianleiwu) [reply docs/ContribOperators.md:2905]: will fix in next pull request

File: onnxruntime/contrib_ops/cuda/bert/packed_multihead_attention.cc (hunk 1 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/bert/packed_multihead_attention.cc:99]: Not needed since we've checked query_dims is either 2 or 4 in the code above.

**Pull request merged**